src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestRunConfigurationProducer.kt

```diff
@@ -83,3 +109,87 @@ private class CargoPackageTestConfig(
     override fun cargoCommandLine(): CargoCommandLine =
         CargoCommandLine.forPackage(cargoPackage, commandName)
 }
+
+class DocTestContext(val owner: RsQualifiedNamedElement, val lineNumber: Int, val isIgnored: Boolean)
+
+// If we encounter one of these attributes, it's probably a doctest.
+// If we encounter something else, it's probably a code block written in some other language.
+// https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html
+private val KNOWN_DOCTEST_ATTRIBUTES = setOf(
+    "compile_fail",
+    "ignore",
+    "rust",
+    "should_panic",
+    "no_run",
+    "edition2015",
+    "edition2018",
+    "edition2021"
+)
+
+fun RsDocCodeFence.getDoctestCtx(): DocTestContext? {
+    val owner = containingDoc.owner as? RsQualifiedNamedElement ?: return null
+
+    val containingFile = originalElement.containingFile
+    val project = containingFile.project
+    val psiDocumentManager = PsiDocumentManager.getInstance(project)
+    val document = psiDocumentManager.getDocument(containingFile) ?: return null
+    val textOffset = originalElement.startOffset
+
+    // Cargo uses 1-based line numbers
+    val lineNumber = document.getLineNumber(textOffset) + 1
+
+    var text = lang?.text?.trim().orEmpty()
+    // Ignore test line marker comments in tests
+    if (isUnitTestMode && "// -" in text) {
+        text = text.substring(0, text.indexOf("// -")).trim()
+    }
+
+    val tags = text.split(",").map { it.trim() }.filter { it.isNotEmpty() }
+    if (tags.isNotEmpty() && !KNOWN_DOCTEST_ATTRIBUTES.contains(tags.first())) return null
```

Comment:
Oh, I had no idea this even existed! Thanks. It would be nice to also reuse `RUST_LANG_ALIASES`, but it doesn't contain `compile_fail`. And it seems that somehow marking compile_fail as being special or handling it out of band would increase the complexity quite a lot.

Btw, how can the unbalanced code fence scenario happen? :thinking:
If we have something like this:
```rust
/// ```
/// let a = 5;
///
/// ```
/// let b = 6;
/// ```
```
There's one fence which is balanced (terminated), and a second one that has no end marker.

Reply:
About unbalanced case:
```rust
/// Some useful docs here
/// ```
/// let a = 1
// This comment (or just en empty like) breaks the doc comment and hence the code fence too
/// let b = a;
/// ```
/// This text was supposed to be just markdown test, but our plugin parses is as a code fence.
/// But `hasUnbalancedCodeFencesBefore` returns `true` for this code fence
/// ```
/// let c = 1; // Also, the plugin parses this code like a markdown text
/// ``` 
```

Reply:
Thanks, I was wondering how to "break" the doctest, a normal comment does it, got it.

Reply:
Another case is a doctest split between outer and inner docs (`///` and `//!`)

Reply:
> My estimate is that in most cases compile_fail doctests fail because of invalid semantics, not invalid syntax, so maybe we could also highlight compile_fail doctests?

I'd want to highlight them without error highlighting. I know there is a hacky way to do that (via `HighlightErrorFilter`, like `FrankensteinErrorFilter`), but I didn't dig deep into it

Reply:
Oh, `HighlightErrorFilter` only suppresses syntax errors. Well, we can skip other errors without a special API like we do with cfg-disabled elements

Reply:
P.S. I don't like these hacks with user data stored in an injected file that `FrankensteinErrorFilter` uses (because of usage of always deprecated `InjectedLanguageUtil`). 
There is a better solution - get `RsDocCodeFence` by an injected file (like we do in `VirtualFile.isDoctestInjection`), then obtain `DoctestInfo` (via `RsDocCodeFence.doctestInfo()`) that (will) store the fact that the doctest is `compile_fail`

Reply:
I suppose that this is a bit orthogonal to this PR, right? I suggest that we keep the current duplication of attributes and later we can fix `compile_fail` and then unify the list of known attributes.

Reply:
Yep